tlc-pack/relax
# [REFACTOR][ARCH] Introduce StructInfo M0

This is the first PR that introduces StructInfo as discussed in https://github.com/tlc-pack/relax/issues/293.

The main list of changes include:

- Introduce StructInfo data structures
- Set of analysis functions that helps to do struct info derivation such as lca, erase_to_well_defined and base analysis 
- Test cases for the analysis functions
- Update printer/parser to handle struct info
- Update normalization/mutator to propagate struct info correct

As a first stab, we aim to bring in minimum best-effort derivation that mostly replicates the previous behavior. We will then send followup PRs to continue bring things to centered around struct info.

This PR is intentionally not meant as a perfect implementation that have all the smart deduction tricks. But a robust one that does as good/better than the existing approach while keeping things as consistent as possible. This helps to bring clarities to future refactors(listed in tracking issue metioned below). Right now, the constructor interface of Var, Function are purposely kept inact and accepts Shape/Type, we then derive struct info from the shape/type. The propagation function calls into UpdateStructInfo, which also automatically updates the shape/type that is consistent with the previous behavior.

All the tests passed locally, with a three exceptions in printer, which we disabled. This is mainly because StructInfo deduction brings more powerful result that returns `Tensor("n * m")`, which we still need to handle in parser and printer. They will be fixed in followup PRs and won't affect our current developments.

We note that the normalization and deduction logic becomes cleaner in the refactor, and as a result motivates us to move to this new state to facilitate future refactors, and we will continue bring set of following PRs for further consolidation
